### PR TITLE
New context API

### PR DESCRIPTION
**Pull request by acdlite**
Introduces a declarative context API that propagates updates even when `shouldComponentUpdate` returns false.

Corresponding RFC: https://github.com/reactjs/rfcs/pull/2

We'll land this behind a feature flag and try it out on a Facebook product (also behind a feature flag) to get some real-world performance data. My plan is to build the React Redux API on top of it, find our largest product that uses Redux, swap out the implementation, and see what happens.

TODO:

- [x] Should be able to stop scanning when we hit a matching provider, then continue the scan later. Already doing this for matching consumers.
- [x] Use ReactElement as child type to save on bundle size.
- [x] Fuzz test.
- [x] Should we use `Object.is` to compare context values, or is `===` enough? `shallowEqual` uses `Object.is`, but we use `===` everywhere else.
- [x] Unwind context stack when interrupted (error, high pri update).
- [x] SSR
- [x] `React.Children`
- [x] Put this behind a feature flag.

### CODE REVIEW

**Comment by gaearon** [comment]
More TODOs:

* SSR support
* React.Children should treat it as a valid type (see the switch in it)

**Comment by gaearon** [comment]
Do you think it's worth copy pasting all/most old tests using context and porting them to the new API?

File: packages/react-reconciler/src/ReactFiberBeginWork.js [hunk 1]
```diff
@@ -606,6 +611,188 @@ export default function<T, P, I, TI, HI, PI, C, CC, CX, PL>(
...
+              } else {
+                // Neither alternate was updated, which means the rest of the
+                // ancestor path already has sufficient priority.
+                break;
+              }
+              node = node.return;
+            }
+            // Don't scan deeper than a matching consumer. When we render the
+            // consumer, we'll continue scanning from that point. This way the
+            // scanning work is time-sliced.
```
Comment (iamdustan) [inline packages/react-reconciler/src/ReactFiberBeginWork.js]: does this imply that the internal subscriptions from react-redux (https://github.com/reactjs/react-redux/blob/master/src/utils/Subscription.js#L58) will be handled by React itself then?

Comment (gaearon) [reply packages/react-reconciler/src/ReactFiberBeginWork.js]: That's the plan

**Comment by jquense** [comment]
we should ping the enzyme folks as well, they'll most likely need to update the traversal logic. @aweary

**Comment by gaearon** [comment]
Why does this affect traversal?

**Comment by jquense** [comment]
ah maybe it doesn't :p it looked like it was adding new `tag` sort. I maybe have just been reading the PR wrong  ignore me!

**Comment by gaearon** [comment]
Ooh I see what you mean. I think @acdlite decided against that though and will keep it an element, overloading `.type` instead. But it's a good point to consider.

**Comment by acdlite** [comment]
@jquense @gaearon The child returned from render will be a `ReactElement`, but it's still using a special Fiber `tag` internally.

**Comment by acdlite** [comment]
> Do you think it's worth copy pasting all/most old tests using context and porting them to the new API?

Eventually yeah. Semantically there's no reason it shouldn't. The biggest open question is performance, so before we worry about upgrade paths we need to find an implementation that works at scale. So I'd rather land this behind a feature flag and try this out on www.

**Comment by iamdustan** [comment]
Fwiw I'd be happy to test this out at Webflow, too, and share any perf data.

File: packages/react-dom/src/server/ReactPartialRenderer.js [hunk 2]
```diff
@@ -116,6 +123,36 @@ if (__DEV__) {
   };
 }
 
+// Context (new API)
+let providerStack: Array<ReactProvider<mixed>> = []; // Stack of provider objects
```
Comment (gaearon) [inline packages/react-dom/src/server/ReactPartialRenderer.js]: Is this reentrant?

Imagine `renderToString(<App />)` where `App` calls `renderToString()` on another tree.

**Comment by gaearon** [comment]
Silly GitHub has hidden my comment. Please don't ignore it. :-)

File: packages/react-dom/src/server/ReactPartialRenderer.js (hunk 2 above)
Comment (acdlite) [reply packages/react-dom/src/server/ReactPartialRenderer.js]: Ooh, yeah, I'll put these on the instance and add a test

**Comment by acdlite** [comment]
Updated with test for reentrant server renders

File: packages/react-dom/src/server/ReactPartialRenderer.js [hunk 3]
```diff
@@ -187,7 +194,7 @@ function warnNoop(
     const constructor = publicInstance.constructor;
     const componentName =
       (constructor && getComponentName(constructor)) || 'ReactClass';
-    const warningKey = `${componentName}.${callerName}`;
```
Comment (acdlite) [inline packages/react-dom/src/server/ReactPartialRenderer.js]: This change is because the previous version was screwing up my editor's syntax highlighting 😆

File: packages/react-reconciler/src/ReactFiberBeginWork.js [hunk 4]
```diff
@@ -605,6 +610,222 @@ export default function<T, P, I, TI, HI, PI, C, CC, CX, PL>(
...
+        // Subscribe to all changes by default
+        bits = MAX_SIGNED_32_BIT_INT;
+      }
+      workInProgress.stateNode = bits;
+
+      if (hasLegacyContextChanged()) {
+        // Normally we can bail out on props equality but if context has changed
+        // we don't do the bailout and we have to reuse existing props instead.
+      } else if (newProps === oldProps && changedBits === 0) {
+        return bailoutOnAlreadyFinishedWork(current, workInProgress);
```
Comment (sebmarkbage) [inline packages/react-reconciler/src/ReactFiberBeginWork.js]: Why should we ever bail out here? Can't we just let this be handled by the next level? Seems like you'd normally always rerender here if you get here. Can simplify this and avoid the extra working managing and storing previous props.

File: packages/react-reconciler/src/ReactFiberBeginWork.js [hunk 5]
```diff
@@ -605,6 +610,222 @@ export default function<T, P, I, TI, HI, PI, C, CC, CX, PL>(
...
+
+      const newProps = workInProgress.pendingProps;
+      const oldProps = workInProgress.memoizedProps;
+
+      // Get the nearest ancestor provider.
+      const providerFiber: Fiber | null = context.currentProvider;
+
+      let newValue;
+      let changedBits;
+      if (providerFiber === null) {
```
Comment (sebmarkbage) [inline packages/react-reconciler/src/ReactFiberBeginWork.js]: This is a hot path and we want these to be able to be sprinkled all over the app so they must be cheap. We also expect the one to many ratio to be big. So whatever work we can do in the provider is worth it.

This whole things could just be:

```js
let newValue = context.currentValue;
let changedBits = context.changedBits;
```

All we need is two new fields on the context type and to set them in the provider (and to reset them when we clear).

File: packages/react-reconciler/src/ReactFiberExpirationTime.js [hunk 6]
```diff
@@ -7,12 +7,14 @@
...
 
+import MAX_SIGNED_32_BIT_INT from './maxSigned32BitInt';
+
 // TODO: Use an opaque type once ESLint et al support the syntax
 export type ExpirationTime = number;
 
 export const NoWork = 0;
 export const Sync = 1;
-export const Never = 2147483647; // Max int32: Math.pow(2, 31) - 1
+export const Never = MAX_SIGNED_32_BIT_INT;
```
Comment (sebmarkbage) [inline packages/react-reconciler/src/ReactFiberExpirationTime.js]: I think we might actually want this to be one bit less since V8 stores signed 31-bit integers as a special path.

File: packages/react-reconciler/src/ReactFiberBeginWork.js [hunk 7]
```diff
@@ -605,6 +610,222 @@ export default function<T, P, I, TI, HI, PI, C, CC, CX, PL>(
...
+      const newValue = newProps.value;
+
+      let changedBits: number;
+      if (oldProps === null) {
+        // Initial render
+        changedBits = MAX_SIGNED_32_BIT_INT;
+      } else {
+        const oldValue = oldProps.value;
+        // Use Object.is to compare the new context value to the old value.
+        if (!is(newValue, oldValue)) {
```
Comment (sebmarkbage) [inline packages/react-reconciler/src/ReactFiberBeginWork.js]: Inline or drop?

**Comment by acdlite** [comment]
More TODOS:

- [ ] Detect concurrent multiple renderers and warn in dev

File: packages/react-reconciler/src/ReactFiberBeginWork.js [hunk 8]
```diff
@@ -605,6 +609,215 @@ export default function<T, P, I, TI, HI, PI, C, CC, CX, PL>(
...
+          // No change.
+          changedBits = 0;
+        } else {
+          changedBits =
+            context.calculateChangedBits !== null
+              ? context.calculateChangedBits(oldValue, newValue)
+              : MAX_SIGNED_31_BIT_INT;
+          if (__DEV__) {
+            warning(
+              (changedBits & MAX_SIGNED_31_BIT_INT) === changedBits,
```
Comment (sophiebits) [inline packages/react-reconciler/src/ReactFiberBeginWork.js]: Why not 32? (changedBits | 0) === changedBits. Fine that the bit field is "negative".

Comment (Kovensky) [reply packages/react-reconciler/src/ReactFiberBeginWork.js]: IIUC, it would cause v8 to store the field in a `double`, even if you flag it as an `int` through the `|0`s.

Comment (acdlite) [reply packages/react-reconciler/src/ReactFiberBeginWork.js]: Yeah that’s the reason. Didn’t know this until @sebmarkbage pointed it out. https://github.com/facebook/react/pull/11818#discussion_r157120923

File: packages/react-reconciler/src/ReactFiberBeginWork.js [hunk 9]
```diff
@@ -605,6 +609,215 @@ export default function<T, P, I, TI, HI, PI, C, CC, CX, PL>(
...
+
+      const newProps = workInProgress.pendingProps;
+      const oldProps = workInProgress.memoizedProps;
+
+      if (hasLegacyContextChanged()) {
+        // Normally we can bail out on props equality but if context has changed
+        // we don't do the bailout and we have to reuse existing props instead.
+      } else if (oldProps === newProps) {
+        pushProvider(workInProgress);
+        workInProgress.stateNode = 0;
```
[the remark this answers is not in the document]
Comment (sebmarkbage) [reply packages/react-reconciler/src/ReactFiberBeginWork.js]: Can you also explain what's going on here? I take it we need to remember this for later but why? Does this have something to do with resuming?

[the remark this answers is not in the document]
Comment (acdlite) [reply packages/react-reconciler/src/ReactFiberBeginWork.js]: The Provider's `stateNode` represents the bits that changed during this render. Setting this to 0 tells the child context that nothing in the provider has changed. I'll add a comment.

**Comment by sebmarkbage** [comment]
Isn’t it sufficient to store the changed bits this render on the context stack instead of the Fiber?

**Comment by acdlite** [comment]
@sebmarkbage You also need to set it when you pop the stack. https://github.com/facebook/react/blob/master/packages/react-reconciler/src/ReactFiberNewContext.js#L55. Could recompute using the `pendingProps` but then you have duplicated logic. Thought it made more sense to cache on the `stateNode`.

**Comment by acdlite** [comment]
It's to avoid doing this work more than once: https://github.com/facebook/react/blob/master/packages/react-reconciler/src/ReactFiberBeginWork.js#L783-L807

**Comment by chenglou** [comment]
Sorry if it's a bit late to chime in. Thinking of supporting this version of the context in ReasonReact. Looks good so far! Though I'd really wish the api was vanilla function based, e.g. `ThemeContext.provide(val, comp)` and `ThemeContext.consume(val => comp)`.

The justification (and this is a slightly selfish one) is that a component's not a function in RR. It's a module/functor, so composition becomes extremely hard. They're barely first-class. So with the current incarnation we'll have to wrap it using some unsafe mechanisms + exposing a function anyway. Would be nice to avoid this. But in general I'm imagining that even in JS folks might start composing/abstracting over context (not saying it's desired. Just that people will), so starting with a function primitive seems a bit more predictable for the eventual composition.

File: packages/react-dom/src/__tests__/ReactServerRendering-test.internal.js [hunk 10]
```diff
@@ -384,6 +387,99 @@ describe('ReactDOMServer', () => {
...
+        );
+      }
+
+      const markup = ReactDOMServer.renderToString(
+        <App value={1} reentrant={true} />,
+      );
+      // Extract the numbers rendered by the consumers
+      const results = markup.match(/\d+/g).map(Number);
+      const reentrantResults = reentrantMarkup.match(/\d+/g).map(Number);
+      expect(results).toEqual([2, 1, 3, 1]);
```
Comment (NE-SmallTown) [inline packages/react-dom/src/__tests__/ReactServerRendering-test.internal.js]: @acdlite Could I ask that why it's `[2, 1, 3, 1]` rather than `[2, 1, 3, 1, 2, 1, 3, 1]`?

I think the markup shoud be 
```
<Provider value={props.value}>
  // created by Reentrant
  <Provider value={props.value}>
    {false}
    <Provider value={2}>
      <Consumer />
    </Provider>
    <Indirection>
      <Indirection>
        <Consumer />
        <Provider value={3}>
          <Consumer />
        </Provider>
      </Indirection>
    </Indirection>
    <Consumer />
  </Provider>
  // created by APP
  <Provider value={2}>
    <Consumer />
  </Provider>
  <Indirection>
    <Indirection>
      <Consumer />
      <Provider value={3}>
        <Consumer />
      </Provider>
    </Indirection>
  </Indirection>
  <Consumer />
</Provider>;
```

If I'm wrong, please let me know, thanks!

**Comment by kentcdodds** [comment]
To be clear, _one_ way that you could use @chenglou's suggestion would be like:

```javascript
const ThemeContext = React.createContext('light')
class ThemeProvider extends React.Component {
  state = {theme: 'light'}
  render() {
    return ThemeContext.provide(this.state.theme, this.props.children)
  }
}

function ThemeConsumer({children}) {
  return ThemeContext.consume(children)
}

class App extends React.Component {
  render() {
    <ThemeProvider>
      <ThemeConsumer>{val => <div>{val}</div>}</ThemeConsumer>
    </ThemeProvider>
  }
}
```

So long as that's possible I'm totally cool with it :+1: (FWIW)

**Comment by acdlite** [comment]
@chenglou @kentcdodds That's actually how this PR currently implements it 🗡 

https://github.com/acdlite/react/blob/0276eb5c5b27ea6ac251c42bd83211c7d486ab16/packages/shared/ReactContext.js#L45-L74

I'm expecting we'll do additional bikeshedding before the release. We could easily support both by attaching `Provider` and `Consumer` properties to that object. And also I guess I'd rename `render` -> `children`.

**Comment by aulisius** [comment]
My 2 cents, I think keeping both render and children might be helpful. It's a very subjective choice.

**Comment by kentcdodds** [comment]
On the subject of render vs children. I just happened to ask this poll yesterday: https://twitter.com/kentcdodds/status/956017113994596352

screen shot 2018-01-24 at 11 29 06 pm

Over 1k responses and the results are fairly even between render (a little ahead) and children.

I personally think it would be great if React core took a stand one way or the other and not include both. It would reduce confusion for the community and hopefully encourage devs to just choose one and stick with it.

**Comment by Andarist** [comment]
A single question (I understand this was discussed under the RFC, but couldnt find a final stance on this) - if I want to use context in lifecycles it's best to consume it and pass as prop to a children which want to use it (which often means introducing intermediary component), right?

**Comment by TrySound** [comment]
@Andarist Yes, there's not third source of truth anymore.

**Comment by drcmda** [comment]
Why doesn't this use a HOC like so many established packages in the react eco system? There are developers on Twitter that have argued against it, but their word stands against the established eco system. I'm a little bit scared of the verbosity that's going to hit, especially with multiple receivers.

I guess it's too late now, but something like the following would have been very familiar to anyone using redux/mobx/etc.

```js
import { consumer } from 'react'

@consumer(
    [SessionContext, ViewContext], // 1. Pass providers
    ([session, view]) => ({ session, camera: view.camera }) // 2. Map provider to props
)
class Csys extends React.Component {
    onClick = () => this.props.session.onClick()
    render() {
        return (
            <Canvas className="csys">
                <View up={this.props.camera.up} session={this.props.session} onClick={this.onClick}/>
            </Canvas>
        )
    }
}
```

Would it really mean we have to nest receivers and pass context up to class methods to access context in them?  

```js
class Csys extends React.Component {
    onClick = () => this.context.session.onClick() // How do i get context in here?
    render() {
        return (
            <SessionContext.Consumer>
                {session => (
                    <ViewContext.Consumer> // Can this be correct? What if i need 10 context props ...
                        {session => (
                            <Canvas className="csys">
                                <View up={view.camera.up} session={session} onClick={this.onClick}/>
                            </Canvas>
                        )}
                    </ViewContext.Consumer>
                )}
            </SessionContext.Consumer>
        )
    }
}
```

**Comment by gaearon** [comment]
“Why not HOC” has been asked and answered several times in the RFC discussion. Please check it out. 

https://github.com/reactjs/rfcs/pull/2

TLDR: because HOC is strictly less powerful and you can always create a HOC on top of this API (but not vice versa).

**Comment by rossipedia** [comment]
Adding my vote for an explicit `render` prop rather than re-using the already overloaded and implicit `children` prop.

Ideally, things should have a single purpose. Children can already be used (and abused) for wildly different things, I don't think adding another thing into the mix is going to help here. 

```js
const App = () => <TrackMouse>{pos => <Page mousePos={pos} />}</TrackMouse>;
```
This doesn't tell me explicitly what the children function is going to be used for. Yes, it's a safe assumption that it will be used for rendering, but I have to _infer_ the intent here. I realize it's an established convention, but it's an _implicit_ one, and children can have other purposes.

Whereas

```js
const App = () => <TrackMouse render={pos => <Page mousePos={pos} />} />;
```

gives an explicit name to the intent. The function is being assigned to a prop with an explicit name `render`, which gives a much stronger signifier of the intent of the function. 

Yes, the component author can still do whatever they want with that prop, but assigning something a name is a much stronger indicator of that thing's use, and is also an established convention across almost every programming language.

**Comment by TrySound** [comment]
@rossipedia I agree. However nested render props are really unreadable. Pair around children makes code much cleaner. Also types allows us to not care about semantics here.

**Comment by gaearon** [comment]
I'd note we're well-aware of the arguments for and against "children" prop naming, they've been rehashed for dozens of times on RFCs, Twitter, and everywhere else :-)

**Comment by acdlite** [comment]
@chenglou 

> But in general, I'm imagining that even in JS, folks might start composing/abstracting over context (not saying it's desired. Just that people will), so starting with a function primitive seems a bit more predictable for the eventual compositions in userland (and if they're any good, it'd be easier to expose context helpers if context's using functions). Also, way simpler semantics.

Just realized I may have misinterpreted what you were asking for.

Even if we go with the function forms, (`ThemeContext.provide(theme, children)`, `Theme.ContextConsume(theme => children)`) you can't call them from anywhere you like. The return value of those functions are React children that must be rendered. In other words, the render prop is lazily evaluated. You can't call this in a lifecycle or event handler and expect to get a callback with the current context. That model would have more complicated semantics, not simpler. The potential for confusion around the API is one of the main reasons we're hesitant about supporting it.

Maybe this isn't what you actually asking for, just wanted to clarify in case anyone reading along is mistaken. The core semantics of this feature have already been decided; what's left to bikeshed are superficial API details.

File: packages/react-dom/src/server/ReactPartialRenderer.js (hunk 3 above)
Comment (batjko) [reply packages/react-dom/src/server/ReactPartialRenderer.js]: Sounds like the worst possible reason for a change in a project with more than one developer.

Comment (acdlite) [reply packages/react-dom/src/server/ReactPartialRenderer.js]: I can think of worse ones

**Comment by gaearon** [comment]
>Even if we go with the function forms, (ThemeContext.provide(theme, children), ThemeContext.consume(theme => children)) you can't call them from anywhere you like. The return value of those functions are React children that must be rendered. 

IMO that's actually a compelling argument for leaving them as JSX-first. Because otherwise it's too east to think `consume` is eagerly evaluated.

**Comment by chenglou** [comment]
@acdlite the return value will be a react element which can't be misused on the Reason side (aka it's not unit/void)

**Comment by mjackson** [comment]
My vote would be to keep the component API. Granted, I haven't used Reason React very much, so I'm not familiar with @chenglou's constraints. But there's no reason it can't be done using plain components in JavaScript-land.

**Comment by mjackson** [comment]
Hey, sorry for commenting on this already-merged PR. But I thought this would be the best place to leave some feedback since it's in context with @chenglou's comments. (No pun intended)

The main reason I dislike the `provide`/`consume` API for end users is that it puts the responsibility on the user to know when to call those functions. The API isn't declarative anymore.

The beauty of the component-based approach is that we can share lifecycle boilerplate code with end users so they don't have to know *when* to call `provide`/`consume`.

When @acdlite first proposed the new context API (component based), I was hoping I could just deprecate and get rid of `react-broadcast`. But if the final API is a pair of functions, I'll probably keep it around and just call `provide`/`consume` behind the scenes as needed instead of using the old context API. Maybe that's ok?

Basically what I'm saying is that for end users, nothing has really changed, except perhaps the fact that this API is now blessed by core and doesn't have the scary warning message in the docs.

**Comment by jquense** [comment]
Idk the `provide/consume` API is no different feeling than `createElement(Consumer, null, render)` plus given that the portal API is `createPortal` not `<Portal />` there is a nice symmetry between those API's for creating "special" elements

File: packages/react-reconciler/src/ReactFiberBeginWork.js [hunk 11]
```diff
@@ -657,6 +661,215 @@ export default function<T, P, I, TI, HI, PI, C, CC, CX, PL>(
...
+        observedBits = MAX_SIGNED_31_BIT_INT;
+      }
+      workInProgress.stateNode = observedBits;
+
+      const newChildren = newProps.render(newValue);
+      reconcileChildren(current, workInProgress, newChildren);
+      return workInProgress.child;
+    } else {
+      return null;
+    }
```
Comment (burimshala) [inline packages/react-reconciler/src/ReactFiberBeginWork.js]: I think the else in the end is unnecessary. You can simply return null. Also I'd suggest to refactor this into another functions because of readability.

File: packages/react-dom/src/server/ReactPartialRenderer.js [hunk 12]
```diff
@@ -642,6 +654,39 @@ class ReactDOMServerRenderer {
...
+    if (this.providerIndex < 0) {
+      context.currentValue = context.defaultValue;
+    } else {
+      // We assume this type is correct because of the index check above.
+      const previousProvider: ReactProvider<any> = (this.providerStack[
+        this.providerIndex
+      ]: any);
+      context.currentValue = previousProvider.props.value;
+    }
   }
```
Comment (burimshala) [inline packages/react-dom/src/server/ReactPartialRenderer.js]: Maybe there is a way to refactor this so you don't have almost the same implementation as in https://github.com/acdlite/react/blob/0276eb5c5b27ea6ac251c42bd83211c7d486ab16/packages/react-reconciler/src/ReactFiberNewContext.js

**Comment by mjackson** [comment]
@jquense Ah, I missed that `provide` and `consume` actually return React elements. That makes sense. Agree the symmetry is nice for creating "special elements". 👍

**Comment by gaearon** [comment]
>I missed that provide and consume actually return React elements

That was the point I was trying to make in https://github.com/ReactTraining/react-broadcast/pull/48#issuecomment-361298855, I guess not very successfully 😛 

>plus given that the portal API is createPortal not <Portal /> there is a nice symmetry between those API's for creating "special" elements

The `createPortal` case is not exactly intentional; we decided to remove `unstable_` prefix at the last minute and thus it ended up in an “underdesigned” state. See https://github.com/facebook/react/issues/11586#issuecomment-345687206.

File: packages/react-reconciler/src/ReactFiberBeginWork.js (hunk 11 above)
Comment (NE-SmallTown) [reply packages/react-reconciler/src/ReactFiberBeginWork.js]: > I think the else in the end is unnecessary

This is just another code style, IMO, current implementation is for more readability

**Comment by MrLoh** [comment]
I’m really glad that this has landed. I’ve recently become a bit concerned with context based subscriptions in react native though and was wondering wether someone can shed some light on it for me. 

A (maybe the) performance issue in react native from my experience has to do with the fact that navigation on mobile works completely different. No matter wether you use react-navigation, react-native-navigation (or even react-router). Unlike on the web, on the native side not only the active view/route is rendered, but a whole stack of views/routes or even several stacks on different bottom nav tabs are rebndered. 

This has huge implications for every subscription based library since it can lead to situations where a state change in let’s say a redux-store or apollo-cache leads to rerendering of several views that are currently not even visible. The reactive nature of React can become a bit problematic here as valuable CPU time is wasted on rendering things that are not currently visible.  (Additionally older devices with slow CPUs, are very common especially outside the US — I’m based in Germany and even here iPhone 5s and 5 year old androids are very common). 

This is to some degree of course a React Native issue but it has a lot to do with Reacts nature and I read very little about it and many libraries are not very familiar with React Native specific issues. Context is one big peace in that puzzle as subscriptions are usually the culprits in unnecessary rerenders. The best way around this at the moment are custom shouldUpdate functions that prevent rerendering of background views until they become active again, but this can’t prevent any logic that runs inside the subscriber HOCs or Consumers, as they escape the normal render cycle through their subscriptions. 

This new context API seems to try to address the subscriber problematic, but I don’t understand yet wether it can also help to prevent triggering unnecessary subscription updates in any way. Maybe someone can illuminate me or has some thoughts on this particular React Native problem.

**Comment by gaearon** [comment]
The bitmask part of the API might help for that. Wait for the blog post and documentation to learn the API in more detail.

**Comment by miraage** [comment]
Hello @acdlite @gaearon . I would like to ask a question regarding this new API. 
I'm using selectors as well as combined selectors via `reselect`. Is it possible to have similar thing for new context API? I'm afraid if it triggers uneccessary renders in case I'll put a POJO as context value.

**Comment by TrySound** [comment]
@miraage Instead of using selectors you save derived data in state. Consider `getDerivedStateFromProps` hook as reducer with `props` instead of `action`. Also unnecessary renders won't happen since changing props always triggers render (unless you use `shouldComponentUpdate`).

**Comment by drcmda** [comment]
@miraage all this is done in react-contextual: https://github.com/drcmda/react-contextual

It prevents the provider from re-rendering, it maps context values to props (so that if you use a PureComponent it uses shallow equal) and it is readily compatible with any memoization strategy, for instance reselect.

File: packages/react-dom/src/server/ReactPartialRenderer.js [hunk 13]
```diff
@@ -764,8 +824,62 @@ class ReactDOMServerRenderer {
...
+            const frame: Frame = {
+              type: provider,
+              domNamespace: parentNamespace,
+              children: nextChildren,
+              childIndex: 0,
+              context: context,
+              footer: '',
+            };
+            if (__DEV__) {
+              ((frame: any): FrameDev).debugElementStack = [];
```
Comment (Yiin-) [inline packages/react-dom/src/server/ReactPartialRenderer.js]: Can someone explain this syntax please?

Comment (iamdustan) [reply packages/react-dom/src/server/ReactPartialRenderer.js]: This is a way to do a relatively unsafe typecast with flow https://flow.org/

```jsx
(variableReference: type);
```

We can’t do the following to cast `frame` to a `FrameDev` because flow already understands it to be of type `Frame`:

```jsx
(frame: FrameDev).debugElementStack = [];
```

By first saying `(frame: any)` we effectively erase the knowledge that it is a `Frame` to be able to say specifically that it is of type `FrameDev`.

Definitely grain of salt the specifics on that regarding how flow internally represents this, but that should give you a high level view of what’s going on.

Comment (Yiin-) [reply packages/react-dom/src/server/ReactPartialRenderer.js]: Thanks!

File: packages/react-reconciler/src/ReactFiberBeginWork.js (hunk 11 above)
Comment (nelsonkonrad) [reply packages/react-reconciler/src/ReactFiberBeginWork.js]: testing

**Comment by PerspectivesLab** [comment]
you guyz can tell how to migrate from 'React.createContext' ??? is there a doc, or anything related ?

**Comment by TrySound** [comment]
Yes, there are docs https://reactjs.org/docs/context.html
